File: backend/routers/meta.py
from fastapi import APIRouter

from backend.config import get_settings

router = APIRouter(tags=["meta"])


@router.get("/api/meta")
def meta() -> dict:
    """Public client configuration — safe to read before login."""
    settings = get_settings()
    return {
        "demo_mode": settings.demo_mode,
        # In demo mode the import screen is hidden and destructive actions
        # (delete, bulk edit, import commit) return 403.
        "import_enabled": not settings.demo_mode,
    }

File: frontend/src/lib/meta.ts
import { useQuery } from "@tanstack/react-query";

import { api } from "./api";

export interface Meta {
  demo_mode: boolean;
  import_enabled: boolean;
}

const FALLBACK: Meta = { demo_mode: false, import_enabled: true };

/**
 * Public instance configuration. Drives demo-mode UI: hiding the import
 * screen and disabling destructive actions. Safe to call before login.
 */
export function useMeta(): Meta {
  const { data } = useQuery({
    queryKey: ["meta"],
    queryFn: () => api<Meta>("/api/meta"),
    staleTime: Infinity,
  });
  return data ?? FALLBACK;
}
